src/terminal_chat_console.cpp

```diff
@@ -0,0 +1,432 @@
+				complete_redraw_needed = true;
+				// This is only used for direct replies from commands
+				// or for lua's print() functionality
+				m_chat_backend.addMessage(L"", evt.evt_msg);
+				break;
+			case CET_TIME_INFO:
+				m_game_time = evt.game_time;
+				m_time_of_day = evt.time;
+		};
+	}
+	while (!m_log_output.queue.empty()) {
+		complete_redraw_needed = true;
+		std::pair<LogLevel, std::string> p = m_log_output.queue.pop_frontNoEx();
+		if (p.first > m_log_level)
+			continue;
+
+		m_chat_backend.addMessage(
+			utf8_to_wide(Logger::getLevelLabel(p.first)),
+			utf8_to_wide(p.second));
+	}
+
+	// handle input
+	if (!m_esc_mode) {
+		handleInput(ch, complete_redraw_needed);
+	} else {
+		switch (ch) {
+			case ERR: // no input
+				break;
+			case 27: // ESC
+				// Toggle ESC mode
+				m_esc_mode = !m_esc_mode;
+				break;
+			case 'L':
+				m_log_level--;
+				m_log_level = MYMAX(m_log_level, LL_NONE + 1); // LL_NONE isn't accessible
+				break;
+			case 'l':
+				m_log_level++;
+				m_log_level = MYMIN(m_log_level, LL_MAX - 1);
+				break;
+		}
+	}
+
+	// was there a resize?
+	int xn, yn;
+	getmaxyx(stdscr, yn, xn);
+	if (xn != m_cols || yn != m_rows) {
+		m_cols = xn;
+		m_rows = yn;
+		reformat_backend(&m_chat_backend, m_rows, m_cols);
+		complete_redraw_needed = true;
+	}
+
+	// draw title
+	move(0, 0);
+	clrtoeol();
+	printw(PROJECT_NAME_C);
+	printw(" ");
+	printw(g_version_hash);
+
+	u32 minutes = m_time_of_day % 1000;
+	u32 hours = (m_time_of_day - minutes) / 1000;
+	minutes = (float)minutes / 1000 * 60;
+
+	if (m_game_time)
+		printw(" | Game %d Time of day %02d:%02d ",
+			m_game_time, hours, minutes);
+
+	// draw text
+	if (complete_redraw_needed)
+		draw_text();
+
+	// draw prompt
+	if (!m_esc_mode) {
+		// normal prompt
+		ChatPrompt& prompt = m_chat_backend.getPrompt();
+		std::string prompt_text = wide_to_utf8(prompt.getVisiblePortion());
+		move(m_rows - 1, 0);
+		clrtoeol();
+		printw("%s", prompt_text.c_str());
+		// Draw cursor
+		s32 cursor_pos = prompt.getVisibleCursorPosition();
+		if (cursor_pos >= 0) {
+			move(m_rows - 1, cursor_pos);
+		}
+	} else {
+		// esc prompt
+		move(m_rows - 1, 0);
+		clrtoeol();
+		printw("[ESC] Toogle ESC mode |"
+			" [CTRL+C] Shut down |"
+			" (L) in-, (l) decrease loglevel %s",
+			Logger::getLevelLabel((LogLevel) m_log_level).c_str());
+	}
+
+	refresh();
+}
+
+void TerminalChatConsole::draw_text()
+{
+	ChatBuffer& buf = m_chat_backend.getConsoleBuffer();
+	for (u32 row = 0; row < buf.getRows(); row++) {
+		move_for_backend(row, 0);
+		clrtoeol();
+		const ChatFormattedLine& line = buf.getFormattedLine(row);
+		if (line.fragments.empty())
+			continue;
+		for (u32 i = 0; i < line.fragments.size(); ++i) {
+			const ChatFormattedFragment& fragment = line.fragments[i];
+			printw("%s", wide_to_utf8(fragment.text).c_str());
+		}
+	}
+}
+
+void TerminalChatConsole::stopAndWaitforThread()
+{
+	clearKillStatus();
+	stop();
+	wait();
+}
+
+TerminalChatConsole g_term_console;
```

Comment:
Don't you think it would be better for visibility to declare globals at the top of the file instead?